### PR TITLE
add UMAP and HDBSCAN params in cluster_ids() method #30

### PR DESCRIPTION
**Pull request by WalidERRAIS**

### CODE REVIEW

File: pykanto/dataset.py [hunk 1]
```diff
@@ -749,7 +749,7 @@ def write_to_json(self) -> None:
         save_to_jsons(self)
 
     @timing
-    def cluster_ids(self, min_sample: int = 10) -> None:
+    def cluster_ids(self, min_sample: int = 10, **kwargs) -> None:
```
Comment (nilomr) [inline pykanto/dataset.py:752]: You would need to pass different kwargs to `umap_reduce` and `hdbscan_cluster`: the current changes pass all kwargs to both functions, which leads to conflicts/errors. See:

- https://gist.github.com/pykong/6a9b020e4df2f56f23e89f15733dbc4b
- https://colabdoge.medium.com/dealing-with-multiple-kwargs-in-python-6babd7be531c
- https://stackoverflow.com/questions/71713742/workaround-for-multiple-parameters-are-not-allowed

File: pykanto/signal/cluster.py [hunk 2]
```diff
@@ -75,7 +74,6 @@ def umap_reduce(
         reducer = cumlUMAP(
             n_neighbors=n_neighbors,
             n_components=n_components,
-            min_dist=min_dist,
```
Comment (nilomr) [inline pykanto/signal/cluster.py:78]: These should be passed as part of the kwargs dict for the same reason as comment in L52-55

**Review by nilomr — CHANGES_REQUESTED** [review]
Hi @WalidERRAIS, thanks a lot for contributing! :smile: 

Had a quick look at your proposed PR #31, see a couple of comments!

File: pykanto/dataset.py [hunk 3]
```diff
@@ -763,10 +763,11 @@ def cluster_ids(self, min_sample: int = 10) -> None:
             min_sample (int): Minimum sample size below which an ID will
                 be skipped. Defaults to 10, but you can reallistically expect
                 good automatic results above ~100.
-
+            kwargs_umap (dict): dictionary of umap params.
+            kwargs_hdbscan (dict): dictionary of hdbscan params.
         """
         df = reduce_and_cluster_parallel(
-            self, min_sample=min_sample, num_cpus=self.parameters.num_cpus
+            self, kwargs_umap=kwargs_umap, kwargs_hdbscan=kwargs_hdbscan, min_sample=min_sample, num_cpus=self.parameters.num_cpus
```
Comment (nilomr) [inline pykanto/dataset.py:770]: Better place the new arguments at the end for backwards compatibility :)

**Review by nilomr — COMMENTED** [review]
Great stuff! Thanks @WalidERRAIS

If you have the time it would be good if you can add unit tests to the PR to make sure this works - if you have never done this/can't I will try to find some time to do it myself once you fix these couple of minor things.

**Comment by nilomr** [comment]
Once you make those changes I think this will be good to go! You can check that it works with a minimal example along the lines of the one below:

```python

from pykanto.utils.paths import pykanto_data
from pykanto.dataset import KantoData
from pykanto.parameters import Parameters

DATASET_ID = "GREAT_TIT"
DIRS = pykanto_data(dataset=DATASET_ID)
# ---------
params = Parameters() 
dataset = KantoData(DIRS, parameters=params, overwrite_dataset=True)

# Segment:
dataset.segment_into_units()

# Plot an example:
dataset.plot(dataset.data.index[0], segmented=True)

# Extract units, project and check with the interactive app:
dataset.get_units()
dataset.cluster_ids(min_sample=5, kwargs_umap={"n_neighbors": 5, "n_components":2, "min_dist": 0.2}, kwargs_hdbscan={"min_cluster_size": 10, "min_samples":1})
dataset.prepare_interactive_data()
dataset.open_label_app()
```

The relevant test that needs to be expanded is https://github.com/waliderrais/pykanto/blob/pr/WalidERRAIS/31/tests/test_dataset.py#L84

**Comment by WalidERRAIS** [comment]
> Great stuff! Thanks @WalidERRAIS
> 
> If you have the time it would be good if you can add unit tests to the PR to make sure this works - if you have never done this/can't I will try to find some time to do it myself once you fix these couple of minor things.

Hi @nilomr Thank you. I learn a lot from you, sorry for previous mistakes I think I missed one commit. I try with a minimal example It works, I never done a unit test but I try to add one I hope it helps you.

**Comment by nilomr** [comment]
> Hi @nilomr Thank you. I learn a lot from you, sorry for previous mistakes I think I missed one commit. I try with a minimal example It works, I never done a unit test but I try to add one I hope it helps you.

@WalidERRAIS I'm having a very hectic week but I'll merge your PR this weekend. I'm just learning as I go myself, thanks for your contribution! :smile:

**Pull request merged**